src/nupic/encoders/Scalar.cpp

```diff
@@ -0,0 +1,248 @@
+/* ---------------------------------------------------------------------
+ * Numenta Platform for Intelligent Computing (NuPIC)
+ * Copyright (C) 2016, Numenta, Inc.  Unless you have an agreement
+ * with Numenta, Inc., for a separate license for this software code, the
+ * following terms and conditions apply:
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Affero Public License version 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
+ * See the GNU Affero Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero Public License
+ * along with this program.  If not, see http://www.gnu.org/licenses.
+ *
+ * http://numenta.org/licenses/
+ * ---------------------------------------------------------------------
+ */
+
+/** @file
+ * Implementations of the ScalarEncoder and PeriodicScalarEncoder
+ */
+
+#include <cstring> // memset
+#include <cmath>
+#include <nupic/encoders/Scalar.hpp>
+#include <nupic/utils/Log.hpp>
+
+namespace nupic
+{
+  static Real64 extractScalar(const ArrayBase & input)
+  {
+    switch (input.getType())
+    {
+    case NTA_BasicType_Byte:
```

Comment:
No, the type isn't known at compile time. Here `getType` is returning an enum. And the `ScalarEncoder` always takes a Real64 (double) but it allows any number type to be in the input, coercing it into a Real64. That logic requires this switch.